proposed/2022/cpm-dotnet-cli-add-package-support.md

```diff
@@ -0,0 +1,134 @@
+
+### 1. The package reference does not exist
+If the package does not already exist, it should be added along with the appropriate package version to `Directory.packages.props`. The package version should either be the latest version or the one specified in the CLI command. Only the package name (not the version) should be added to `<PackageReference>` in the project file.
+
+#### Before `add package` is executed:<br>
+
+The props file: <br>
+```xml
+<Project>
+    <PropertyGroup>
+        <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
+    </PropertyGroup>
+    <ItemGroup>
+    </ItemGroup>
+</Project>
+```
+The .csproj file: <br>
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+    <PropertyGroup>
+        <TargetFramework>net6.0</TargetFramework>
+        <ImplicitUsings>enable</ImplicitUsings>
+        <Nullable>enable</Nullable>
+    </PropertyGroup>
+
+    <ItemGroup>
+    </ItemGroup>
+</Project>
+```
+
+#### After `add package` is executed:<br>
+
+The props file: <br>
+```xml
+<Project>
+    <PropertyGroup>
+        <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
+    </PropertyGroup>
+    <ItemGroup>
+    <PackageVersion Include="Newtonsoft.Json" Version="13.0.1"/>
+    </ItemGroup>
+</Project>
+```
+The .csproj file: <br>
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+    <PropertyGroup>
+        <TargetFramework>net6.0</TargetFramework>
+        <ImplicitUsings>enable</ImplicitUsings>
+        <Nullable>enable</Nullable>
+    </PropertyGroup>
+
+    <ItemGroup>
+        <PackageReference Include="Newtonsoft.Json"/>
+    </ItemGroup>
+</Project>
+```
+
+In case there are multiple `Directory.packages.props` files in the repo, the props file that is closest must be considered.
+
+```xml
+Repository
+ |-- Directory.Packages.props
+ |-- Solution1
+    |-- Directory.Packages.props
+    |-- Project1
+ |-- Solution2
+    |-- Project2
+```
+
+In the above example, the following scenarios are possible:
+1. Project1 will evaluate the Directory.Packages.props file in the Repository\Solution1\ directory.
+2. Project2 will evaluate the Directory.Packages.props file in the Repository\ directory.
+
+***Sourced from https://devblogs.microsoft.com/nuget/introducing-central-package-management/
+
+### 2. The package reference does exist
+If the package already exists in `Directory.packages.props` the version should be updated in `Directory.packages.props`. The package version should either be the latest version or the one specified in the CLI command. The `<PackageReference>` in the project file should not change.
+
+#### Before `add package` is executed:<br>
+```xml
+<Project>
+    <PropertyGroup>
+        <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
+    </PropertyGroup>
+    <ItemGroup>
+    <PackageVersion Include="Newtonsoft.Json" Version="12.0.1"/>
+    </ItemGroup>
+</Project>
+```
+
+#### After `add package` is executed:<br>
```

Comment:
I think this behavior might not be super obvious to the end user. 

If I am already managing a package centrally, it feels like the version that should be installed is the existing one, and ideally not update it. 

Version updates should only happen explicitly or through an update command.

Reply:
We went down this path because `dotnet add package` command currently updates the package version if the package refence already exists. 

If it is not ideal for `dotnet add package` command to update the package version then should we raise a warning instead of updating the version in the `directory.packages.props` file for projects onboarded onto CPM?